editor/plugins/curve_editor_plugin.cpp

```diff
@@ -557,43 +672,60 @@ void CurveEditor::update_view_transform() {
 	_world_to_view = view_trans * world_trans;
 }
 
-Vector2 CurveEditor::get_tangent_view_pos(int i, TangentIndex tangent) const {
+Vector2 CurveEdit::get_tangent_view_pos(int p_index, TangentIndex p_tangent) const {
 	Vector2 dir;
-	if (tangent == TANGENT_LEFT) {
-		dir = -Vector2(1, _curve_ref->get_point_left_tangent(i));
+	if (p_tangent == TANGENT_LEFT) {
+		dir = -Vector2(1, curve->get_point_left_tangent(p_index));
 	} else {
-		dir = Vector2(1, _curve_ref->get_point_right_tangent(i));
+		dir = Vector2(1, curve->get_point_right_tangent(p_index));
+	}
+
+	Vector2 point_pos = curve->get_point_position(p_index);
+	Vector2 point_view_pos = get_view_pos(point_pos);
+	Vector2 control_view_pos = get_view_pos(point_pos + dir);
+
+	Vector2 distance_from_point = tangent_length * (control_view_pos - point_view_pos).normalized();
+	Vector2 tangent_view_pos = point_view_pos + distance_from_point;
+
+	// Limit the tangent control point inside the rect.
+	float fraction_inside = 1.0;
+	if (distance_from_point.x != 0.0) {
+		fraction_inside = MIN(fraction_inside, ((distance_from_point.x > 0 ? get_rect().size.x : 0) - point_view_pos.x) / distance_from_point.x);
+	}
+	if (distance_from_point.y != 0.0) {
+		fraction_inside = MIN(fraction_inside, ((distance_from_point.y > 0 ? get_rect().size.y : 0) - point_view_pos.y) / distance_from_point.y);
 	}
 
-	Vector2 point_pos = get_view_pos(_curve_ref->get_point_position(i));
-	Vector2 control_pos = get_view_pos(_curve_ref->get_point_position(i) + dir);
+	if (fraction_inside < 1.0 && fraction_inside > 0.5) {
+		tangent_view_pos = point_view_pos + distance_from_point * fraction_inside;
+	}
 
-	return point_pos + Math::round(_tangents_length * _gizmo_handle_scale * EDSCALE) * (control_pos - point_pos).normalized();
+	return tangent_view_pos;
 }
 
-Vector2 CurveEditor::get_view_pos(Vector2 world_pos) const {
-	return _world_to_view.xform(world_pos);
+Vector2 CurveEdit::get_view_pos(Vector2 p_world_pos) const {
+	return _world_to_view.xform(p_world_pos);
 }
 
-Vector2 CurveEditor::get_world_pos(Vector2 view_pos) const {
-	return _world_to_view.affine_inverse().xform(view_pos);
+Vector2 CurveEdit::get_world_pos(Vector2 p_view_pos) const {
+	return _world_to_view.affine_inverse().xform(p_view_pos);
 }
 
-// Uses non-baked points, but takes advantage of ordered iteration to be faster
+// Uses non-baked points, but takes advantage of ordered iteration to be faster.
 template <typename T>
-static void plot_curve_accurate(const Curve &curve, float step, T plot_func) {
+static void plot_curve_accurate(const Curve &curve, float step, Vector2 scaling, T plot_func) {
 	if (curve.get_point_count() <= 1) {
-		// Not enough points to make a curve, so it's just a straight line
+		// Not enough points to make a curve, so it's just a straight line.
 		float y = curve.sample(0);
-		plot_func(Vector2(0, y), Vector2(1.f, y), true);
+		plot_func(Vector2(0, y) * scaling + Vector2(0.5, 0), Vector2(1.f, y) * scaling - Vector2(1.5, 0), true);
```

Comment:
For some reason, in that scenario, the curve would poke out a little before x=0 and a little after x=1, which is fixed by these little tweaks to the vectors. Added a comment.

Reply:
Oooh, I think I see. This is basically just moving the points a couple of pixels - I had misinterpreted this originally as doing a lot more :)

I still do wonder if it's an artifact of the added scaling in this PR, because that didn't use to be a problem originally.